### PR TITLE
[metricbeat][mysql] Add SSL support

### LINKED ISSUE #16133: [Metricbeat] Add TLS/SSL configuration options in the MySQL module
**Describe the enhancement:** following-up this [Discuss topic](https://discuss.elastic.co/t/how-to-add-ssl-certificate-for-mysql-connection-for-metricbeat/217851).

The MySQL module configuration for Metricbeat does not provide any configuration / parameters to specify the list of trusted root certificates or other TLS/SSL parameters (e.g. client certificate, client certificate key, etc.). 

It could be confusing to users that [Metricbeat](https://www.elastic.co/guide/en/beats/metricbeat/current/configuration-ssl.html) provides TLS/SSL options/parameters for Elasticsearch, Kibana, or other modules (e.g http) but does not provide these to the MySQL module itself.

As shown in the [Discuss topic](https://discuss.elastic.co/t/how-to-add-ssl-certificate-for-mysql-connection-for-metricbeat/217851), If users decide to connect to a MySQL database with SSL (i.e adding `tls=true` in the host DSN), they would need to download and add the root certificate in the host certificate store where Metricbeat runs. If the root certificate is not present, Metricbeat will thrown an error: `x509: certificate signed by unknown authority`.

**Describe a specific use case for the enhancement or feature:** Given a MySQL database configured with SSL, provide the ability to configure the MySQL module  for Metricbeat with TLS/SSL options/parameters.

### PR DESCRIPTION
**Pull request by ghost**
## Overview

Adds SSL support for following metricsets:
- [x] query
- [x] performance
- [x] status
- [x] galera_status

The `query` metricset uses a different SQL client (`helper/sql`). Others use the client from the module itself.

Added a config struct as part of this change.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

<details><summary><strong>genroot.sh</strong></summary>
<p>

```sh
OPENSSL_ROOT_CA=$1

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl genrsa 2048 > certs/root-ca-key.pem

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl req -new -x509 -nodes -days 3600 \
        -subj "${OPENSSL_ROOT_CA}" \
        -key /certs/root-ca-key.pem -out /certs/root-ca.pem
```

</p>
</details> 

<details><summary><strong>genserver.sh</strong></summary>
<p>

```sh
OPENSSL_SERVER=$1

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl req -newkey rsa:2048 -days 3600 -nodes \
        -subj "${OPENSSL_SERVER}" \
        -keyout /certs/server-key.pem -out /certs/server-req.pem

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl rsa -in /certs/server-key.pem -out /certs/server-key.pem

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl x509 -req -in /certs/server-req.pem -days 3600 \
        -CA /certs/root-ca.pem -CAkey /certs/root-ca-key.pem \
        -set_serial 01 -out /certs/server-cert.pem

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl verify -CAfile /certs/root-ca.pem /certs/server-cert.pem
```

</p>
</details> 

<details><summary><strong>genclient.sh</strong></summary>
<p>

```sh
OPENSSL_CLIENT=$1

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl req -newkey rsa:2048 -days 3600 -nodes \
        -subj "${OPENSSL_CLIENT}" \
        -keyout /certs/client-key.pem -out /certs/client-req.pem

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl rsa -in /certs/client-key.pem -out /certs/client-key.pem

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl x509 -req -in /certs/client-req.pem -days 3600 \
        -CA /certs/root-ca.pem -CAkey /certs/root-ca-key.pem \
        -set_serial 01 -out /certs/client-cert.pem

docker run --rm -v $PWD/certs:/certs -it nginx \
    openssl verify -CAfile /certs/root-ca.pem /certs/client-cert.pem
```

</p>
</details> 

<details><summary><strong>gencerts.sh</strong></summary>
<p>

```sh
mkdir -p certs

OPENSSL_SUBJ="/C=US/ST=California/L=Santa Clara"
OPENSSL_CA="${OPENSSL_SUBJ}/CN=fake-CA"
OPENSSL_SERVER="${OPENSSL_SUBJ}/CN=fake-server"
OPENSSL_CLIENT="${OPENSSL_SUBJ}/CN=fake-client"

sh ./genroot.sh "${OPENSSL_CA}"
sh ./genserver.sh "${OPENSSL_SERVER}"
sh ./genclient.sh "${OPENSSL_CLIENT}"
```

</p>
</details> 

Then run `sh +x gencerts.sh` to generate required certificates.

Update certificates ownership and permissions so the `mysql` container can use them.

```sh
sudo chown :docker certs/*
sudo chmod 755 certs/*
```

<details><summary><strong>docker-compose.up.yml</strong></summary>
<p>

```yaml
version: '3.8'
services:
  mysql:
    image: mysql/mysql-server:8.0.21
    command: [ "mysqld",
                    "--character-set-server=utf8mb4",
                    "--collation-server=utf8mb4_unicode_ci",
                    "--bind-address=0.0.0.0",
                    "--require_secure_transport=ON",
                    "--ssl-ca=/etc/certs/root-ca.pem",
                    "--ssl-cert=/etc/certs/server-cert.pem",
                    "--ssl-key=/etc/certs/server-key.pem",
                    "--default_authentication_plugin=mysql_native_password" ]
    environment:
      MYSQL_ROOT_PASSWORD: "example"
      MYSQL_ROOT_HOST: "%"
    restart: always
    volumes:
      - type: bind
        source: ./certs
        target: /etc/certs/
    ports:
      - 3306:3306
```

</p>
</details> 

## Related issues

- Closes #16133

### CODE REVIEW

File: metricbeat/module/mysql/mysql.go [hunk 1]
```diff
@@ -86,6 +114,10 @@ func ParseDSN(mod mb.Module, host string) (mb.HostData, error) {
 	noCredentialsConfig.User = ""
 	noCredentialsConfig.Passwd = ""
 
+	if c.TLS.IsEnabled() {
+		config.TLSConfig = TLSConfigKey
+	}
```
Comment (shmsr) [inline metricbeat/module/mysql/mysql.go:119]: why is this different?

File: metricbeat/module/mysql/config.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+package mysql
+
+import (
+	"crypto/tls"
+
+	"github.com/elastic/elastic-agent-libs/transport/tlscommon"
+)
+
+type Config struct {
+	Hosts     []string          `config:"hosts"    validate:"required"`
```
Comment (shmsr) [inline metricbeat/module/mysql/config.go:27]: nitpick: get rid of extra space before `validate`

File: metricbeat/module/mysql/mysql.go [hunk 3]
```diff
@@ -38,26 +42,50 @@ func init() {
...
 	return &base, nil
 }
 
+type Metricset struct {
+	mb.BaseMetricSet
+	Config Config
+}
+
+func NewMetricset(base mb.BaseMetricSet) (*Metricset, error) {
+	// Validate that at least one host has been specified.
```
Comment (shmsr) [inline metricbeat/module/mysql/mysql.go:59]: isn't this comment irrelevant?

**Review by shmsr — COMMENTED** [review]
Changes look more or less good only but I feel I need to do another comprehensive review. I'll do it ASAP.

**Comment by shmsr** [comment]
Can you share the minimum steps to test the changes made in "How to test this PR locally" section of the PR description.

File: CHANGELOG.next.asciidoc [hunk 4]
```diff
@@ -203,6 +203,7 @@ Setting environmental variable ELASTIC_NETINFO:false in Elastic Agent pod will d
 - Add a `/inputs/` route to the HTTP monitoring endpoint that exposes metrics for each metricset instance. {pull}36971[36971]
 - Add linux IO metrics to system/process {pull}37213[37213]
 - Add new memory/cgroup metrics to Kibana module {pull}37232[37232]
+- Add mysql module SSL support {pull}37997[37997]
```
Comment (devamanv) [inline CHANGELOG.next.asciidoc:206]:
```suggestion
- Add SSL support to mysql module {pull}37997[37997]
```
 or simply
```suggestion
- Add SSL support  {pull}37997[37997]
```

File: metricbeat/module/mysql/mysql.go (hunk 1 above)
Comment (gpop63) [reply metricbeat/module/mysql/mysql.go:119]: That's the TLS config name being appended to the mysql DSN.

File: metricbeat/module/mysql/config.go (hunk 2 above)
Comment (devamanv) [reply metricbeat/module/mysql/config.go:27]: +1, looks like a tab but should be a space instead. From the official [docs](https://pkg.go.dev/reflect#StructTag)
>By convention, tag strings are a concatenation of optionally space-separated key:"value" pairs.

File: metricbeat/module/mysql/mysql_integration_test.go [hunk 5]
```diff
@@ -31,7 +31,7 @@ import (
 func TestNewDB(t *testing.T) {
 	service := compose.EnsureUp(t, "mysql")
 
-	db, err := NewDB(GetMySQLEnvDSN(service.Host()))
+	db, err := NewDB(GetMySQLEnvDSN(service.Host()), nil)
```
Comment (rdner) [inline metricbeat/module/mysql/mysql_integration_test.go:34]: Should we introduce a test that validates the added SSL support?

**Review by shmsr — COMMENTED** [review]
The changes look good. Are you planning to add integration tests (what Denis also asked)?

**Review by rdner — APPROVED** [review]

**Review by shmsr — APPROVED** [review]

**Pull request merged**